pulser/sequence.py

```diff
@@ -12,32 +12,47 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import annotations
+
 from collections import namedtuple
-from collections.abc import Iterable
+from collections.abc import Callable, Generator, Iterable
 import copy
 from functools import wraps
 from itertools import chain
 import json
+from typing import (Any, cast, NamedTuple,
+                    Optional, Tuple, Union)
+from typing_extensions import Literal
+
 import warnings
 
 import numpy as np
+from numpy.typing import ArrayLike
 
 import pulser
+from pulser._seq_drawer import draw_sequence
+from pulser.channels import Channel
 from pulser.pulse import Pulse
 from pulser.devices import MockDevice
 from pulser.devices._device_datacls import Device
 from pulser.json.coders import PulserEncoder, PulserDecoder
 from pulser.json.utils import obj_to_dict
 from pulser.parametrized import Parametrized, Variable
-from pulser._seq_drawer import draw_sequence
+from pulser.register import Register
+
+QubitId = Union[int, str]
+PROTOCOLS = Literal['min-delay', 'no-delay', 'wait-for-all']
 
 # Auxiliary class to store the information in the schedule
-_TimeSlot = namedtuple('_TimeSlot', ['type', 'ti', 'tf', 'targets'])
+_TimeSlot: NamedTuple('_TimeSlot', [('type', str), ('ti', int), ('tf', int), (
+    'targets', Iterable[QubitId])]) = namedtuple(
+        '_TimeSlot', ['type', 'ti',
+                      'tf', 'targets'])
```

Comment:
New errors appear:

```
pulser\sequence.py:444: error: Unsupported left operand type for & ("Iterable[Union[int, str]]")
pulser\sequence.py:465: error: Argument 1 to "_TimeSlot" has incompatible type "Pulse"; expected "str"
pulser\sequence.py:752: error: Unsupported operand types for - ("None" and "int")
pulser\sequence.py:752: note: Left operand is of type "Optional[int]"
pulser\sequence.py:756: error: Incompatible types in assignment (expression has type "int", variable has type "Union[number[Any], ndarray]")
pulser\sequence.py:757: error: Argument 1 to "validate_duration" of "Channel" has incompatible type "Union[number[Any], ndarray]"; expected "int"
pulser\sequence.py:765: error: Incompatible types in assignment (expression has type "Union[number[Any], Any]", target has type "int")
pulser\sequence.py:766: error: Argument 3 to "_TimeSlot" has incompatible type "Union[number[Any], Any]"; expected "int"
```

But I feel they make sense and have to be adressed.

For instance, for the error on line 752:

```
delta = np.clip(retarget - elapsed, 0, retarget)
```

We type hinted `Channel.retarget_time` as `Optional[int] = None`. Shouldn't we check if it is `None` and, if it is, force 0 in the calculations ?

Reply:
Ok, a few things:
- For `'type'`, it can indeed be `Union[Pulse, str]`
- `targets` is always `Set[QubitId]` (this is `collections.abc.Set`, which is substitutes `typing.AbstractSet`)
- For line 752, the channel has to be "Local", which forces a `retarget_time` to be defined. The parameter is `Optional` because it shouldn't be defined when the channel is "Global". So, in the case you mentioned, it's safe to cast to `int`

Try to address these and we'll see what remains